# AoC Day 17

Jenna Jordan

18 December 2021

## Prompt

--- Day 17: Trick Shot ---

You finally decode the Elves' message. HI, the message says. You continue searching for the sleigh keys.

Ahead of you is what appears to be a large ocean trench. Could the keys have fallen into it? You'd better send a probe to investigate.

The probe launcher on your submarine can fire the probe with any integer velocity in the x (forward) and y (upward, or downward if negative) directions. For example, an initial `x,y` velocity like `0,10` would fire the probe straight up, while an initial velocity like `10,-1` would fire the probe forward at a slight downward angle.

The probe's x,y position starts at 0,0. Then, it will follow some trajectory by moving in steps. On each step, these changes occur in the following order:

- The probe's x position increases by its x velocity.
- The probe's y position increases by its y velocity.
- Due to drag, the probe's x velocity changes by 1 toward the value 0; that is, it decreases by 1 if it is greater than 0, increases by 1 if it is less than 0, or does not change if it is already 0.
- Due to gravity, the probe's y velocity decreases by 1.

For the probe to successfully make it into the trench, the probe must be on some trajectory that causes it to be within a target area after any step. The submarine computer has already calculated this target area (your puzzle input). For example:

`target area: x=20..30, y=-10..-5`

This target area means that you need to find initial x,y velocity values such that after any step, the probe's x position is at least 20 and at most 30, and the probe's y position is at least -10 and at most -5.

Given this target area, one initial velocity that causes the probe to be within the target area after any step is 7,2:

```
.............#....#............
.......#..............#........
...............................
S........................#.....
...............................
...............................
...........................#...
...............................
....................TTTTTTTTTTT
....................TTTTTTTTTTT
....................TTTTTTTT#TT
....................TTTTTTTTTTT
....................TTTTTTTTTTT
....................TTTTTTTTTTT
```

In this diagram, S is the probe's initial position, 0,0. The x coordinate increases to the right, and the y coordinate increases upward. In the bottom right, positions that are within the target area are shown as T. After each step (until the target area is reached), the position of the probe is marked with #. (The bottom-right # is both a position the probe reaches and a position in the target area.)

Another initial velocity that causes the probe to be within the target area after any step is 6,3:

```
...............#..#............
...........#........#..........
...............................
......#..............#.........
...............................
...............................
S....................#.........
...............................
...............................
...............................
.....................#.........
....................TTTTTTTTTTT
....................TTTTTTTTTTT
....................TTTTTTTTTTT
....................TTTTTTTTTTT
....................T#TTTTTTTTT
....................TTTTTTTTTTT
```

Another one is 9,0:

```
S........#.....................
.................#.............
...............................
........................#......
...............................
....................TTTTTTTTTTT
....................TTTTTTTTTT#
....................TTTTTTTTTTT
....................TTTTTTTTTTT
....................TTTTTTTTTTT
....................TTTTTTTTTTT
```

One initial velocity that doesn't cause the probe to be within the target area after any step is 17,-4:

```
S..............................................................
...............................................................
...............................................................
...............................................................
.................#.............................................
....................TTTTTTTTTTT................................
....................TTTTTTTTTTT................................
....................TTTTTTTTTTT................................
....................TTTTTTTTTTT................................
....................TTTTTTTTTTT..#.............................
....................TTTTTTTTTTT................................
...............................................................
...............................................................
...............................................................
...............................................................
................................................#..............
...............................................................
...............................................................
...............................................................
...............................................................
...............................................................
...............................................................
..............................................................#
```

The probe appears to pass through the target area, but is never within it after any step. Instead, it continues down and to the right - only the first few steps are shown.

If you're going to fire a highly scientific probe out of a super cool probe launcher, you might as well do it with style. How high can you make the probe go while still reaching the target area?

In the above example, using an initial velocity of 6,9 is the best you can do, causing the probe to reach a maximum y position of 45. (Any higher initial y velocity causes the probe to overshoot the target area entirely.)

Find the initial velocity that causes the probe to reach the highest y position and still eventually be within the target area after any step. What is the highest y position it reaches on this trajectory?

In [1]:
# get input

with open("inputs/day17.txt") as file:
    input_str = file.read().rstrip()
    
test_input_str = "target area: x=20..30, y=-10..-5"

In [2]:
test_input_str

'target area: x=20..30, y=-10..-5'

In [3]:
import re
import itertools

In [4]:
def get_target_area(input_str):
    numbers = re.search("x=(-?\d+)\.\.(-?\d+), y=(-?\d+)\.\.(-?\d+)", input_str)
    return ((int(numbers.group(1)), int(numbers.group(2))), (int(numbers.group(3)), int(numbers.group(4))))

In [5]:
def step_trajectory(start_position, trajectory):
    x_pos, y_pos = start_position
    x_vel, y_vel = trajectory
    x_pos += x_vel
    y_pos += y_vel
    if x_vel > 0:
        x_vel -= 1
    elif x_vel < 0:
        x_vel += 1
    y_vel -= 1
    
    return (x_pos, y_pos), (x_vel, y_vel) # new position x,y and new velocity x,y

In [6]:
def check_if_in_target_area(position, target_area):
    ((x_min, x_max),(y_min, y_max)) = target_area
    x_pos, y_pos = position
    if (x_min <= x_pos <= x_max) and (y_min <= y_pos <= y_max):
        return True
    else:
        return False

In [7]:
def get_highest_position(positions):
    all_heights = [p[1] for p in positions]
    return max(all_heights)

In [8]:
def check_if_trajectory_hits_target(trajectory, target_area):
    all_positions = []
    target_not_reached = True
    position = (0,0)
    hits_target = False
    while target_not_reached:
        position, trajectory = step_trajectory(position, trajectory)
        all_positions.append(position)
        x_pos, y_pos = position
        ((x_min, x_max),(y_min, y_max)) = target_area
        if check_if_in_target_area(position, target_area):
            hits_target = True
        if (x_pos > x_max) or (y_pos < y_min):
            target_not_reached = False
    max_height = get_highest_position(all_positions)
    return hits_target, max_height

In [9]:
test_target = get_target_area(test_input_str)
test_traj1 = (7,2)
test_traj2 = (6,3)
test_traj3 = (9,0)
test_traj4 = (17,-4)

In [10]:
check_if_trajectory_hits_target(test_traj4, test_target)

(False, -4)

In [11]:
def generate_all_trajectories(x_max, y_min, y_max):
    x_vels = [x for x in range(x_max+1)]
    y_vels = [x for x in range(y_min, y_max+1)]
    return list(itertools.product(x_vels, y_vels))

In [12]:
def choose_best_trajectory(target_area):
    max_height = 0
    best_traj = (0,0)
    ((x_min, x_max),(y_min, y_max)) = target_area
    all_trajectories = generate_all_trajectories(x_max, -(abs(y_min)), abs(y_min))
    for t in all_trajectories:
        hits, height = check_if_trajectory_hits_target(t, target_area)
        if hits:
            if height > max_height:
                max_height = height
                best_traj = t
    return max_height, best_traj

In [13]:
choose_best_trajectory(test_target)

(45, (6, 9))

In [14]:
target = get_target_area(input_str)
choose_best_trajectory(target)

(6903, (22, 117))

### Part 1 Solution

6903 - CORRECT!

--- Part Two ---

Maybe a fancy trick shot isn't the best idea; after all, you only have one probe, so you had better not miss.

To get the best idea of what your options are for launching the probe, you need to find every initial velocity that causes the probe to eventually be within the target area after any step.

In the above example, there are 112 different initial velocity values that meet these criteria:

```
23,-10  25,-9   27,-5   29,-6   22,-6   21,-7   9,0     27,-7   24,-5
25,-7   26,-6   25,-5   6,8     11,-2   20,-5   29,-10  6,3     28,-7
8,0     30,-6   29,-8   20,-10  6,7     6,4     6,1     14,-4   21,-6
26,-10  7,-1    7,7     8,-1    21,-9   6,2     20,-7   30,-10  14,-3
20,-8   13,-2   7,3     28,-8   29,-9   15,-3   22,-5   26,-8   25,-8
25,-6   15,-4   9,-2    15,-2   12,-2   28,-9   12,-3   24,-6   23,-7
25,-10  7,8     11,-3   26,-7   7,1     23,-9   6,0     22,-10  27,-6
8,1     22,-8   13,-4   7,6     28,-6   11,-4   12,-4   26,-9   7,4
24,-10  23,-8   30,-8   7,0     9,-1    10,-1   26,-5   22,-9   6,5
7,5     23,-6   28,-10  10,-2   11,-1   20,-9   14,-2   29,-7   13,-3
23,-5   24,-8   27,-9   30,-7   28,-5   21,-10  7,9     6,6     21,-5
27,-10  7,2     30,-9   21,-8   22,-7   24,-9   20,-6   6,9     29,-5
8,-2    27,-8   30,-5   24,-7
```
test_target
How many distinct initial velocity values cause the probe to be within the target area after any step?

In [15]:
def generate_all_trajectories(x_max, y_min, y_max):
    x_vels = [x for x in range(x_max+1)]
    y_vels = [x for x in range(y_min, y_max+1)]
    return list(itertools.product(x_vels, y_vels))

In [16]:
def get_valid_trajectories(target_area):
    valid_trajectories = []
    ((x_min, x_max),(y_min, y_max)) = target_area
    all_trajectories = generate_all_trajectories(x_max, -(abs(y_min)), abs(y_min))
    for t in all_trajectories:
        hits, height = check_if_trajectory_hits_target(t, target_area)
        if hits:
            valid_trajectories.append(t)
    return valid_trajectories

In [17]:
test_valid_traj = get_valid_trajectories(test_target)
len(test_valid_traj)

112

In [19]:
valid_traj = get_valid_trajectories(target)
len(valid_traj)

2351

### Part 2 Solution

2351 - CORRECT!